### PR TITLE
Some opti

### PR DESCRIPTION
**Pull request by DevNullx64**
I've miss function and call flow that was fully use less after the rewrite (ResetGrad() by exemple)
I've removed the creation of the graph using Visited and TopOSort. The backpropagation itselft flow directly from the loss.Backpropagate() node and internaly call LeftChildren.Backpropagate_() and RightChildren.Backpropagate_() recursively.

I've miss to update the readme in accordance :/

### CODE REVIEW

**Comment by Leonardo16AM** [comment]
Hello, I will review the changes, sorry for the delay. The reason why I did the Toposort is that the operations create an Acyclic Directed Graph, when applying the backpropagation on a node you have to make sure that all the nodes that it can reach have already calculated their gradient, if you transmit them recursively I think that property is not fulfilled.

**Comment by DevNullx64** [comment]
Got it. So, I've currently break the cicle protection. I will rollback this regression asap

**Comment by DevNullx64** [comment]
All my appologies if my question sound like stupid.

If the building of the graph is maintain by the overided operator logic. Do you have of a way to create a cycle ? Currently, Sharp-Grad can handle Value with an already know content or other Value. Prehaps I miss somthing but I can't react a cicle then it's hard to me to know how the ensure this cyclic protections.

**Comment by Leonardo16AM** [comment]
The main problem is not that cycles are created, in fact I think that they should not be created, because it should be directed acyclic graph. I give you an example of the problem:

B=A+A
C=A+A
D=C+B
D.backpropagate()

In this case you will have a dependency graph that will look like a Diamond, something like:

A->B
B->D
A->C
C->D
(These are the directed edges, deleting the repeated ones)


 Now if you backpropagate recursively starting in D, the recursion will go to B, then from B to A, and in A it will need to calculate the gradient, but the recursion has not calculated yet the gradient of C, so it will have a wrong value. Also, can happen that the recursive function goes from D to B , and from B to A, but later it will go from D to C and from C to A, take notice that you are visiting A twice, so in this case it is not a big deal, but it can have an exponential time complexity.

Topological Sort solves those two problems, it gives you an array in which if you have two nodes U and V, and V is a descendant of U in the DAG, it will appear before in the array, so you can apply the backward pass in the order of the Toposort. And the time Complexity will be O(MlogN)  Where M is the number of edges, and de LogN is due to the HashMap

**Comment by DevNullx64** [comment]
Understood. I believe we can close this PR, especially since I've juste removed the two unnecessary functions. Rolling back seems like the best approach.

Regarding your point about the complexity explosions when performing multiple passes on a node in the compute graph, from my perspective, in the worst-case scenario, a node may be computed once per edge. We could establish a rough upper bound on the order of the node with the most edges in computation.

In the next phase outlined in my roadmap, the VRAM bandwidth will likely be the bottleneck. I'm exploring ways to minimize memory consumption. Getting lost in O(X) complexity isn't my primary goal. Instead, I'm focusing on leveraging GPU registers and memory cache more effectively. The objective is to ensure that computation time is generally shorter than loading time. I prefer the GPU to spend cycles recomputing rather than waiting for data from VRAM. In terms of computational accuracy, do you know if accumulating the gradient and then performing backward propagation once is computationally equivalent to performing multiple backward passes on the same node?

FYI: My English is not bad nor very good, so I'm using GPT to rephrase my sentences ;)

**Comment by Leonardo16AM** [comment]
Ok, rolling back the PR seems to be the best.

Regarding the time complexity of the recursion, i think it can be up to 2^(N/2). For example:

image

This is due that you are not marking the nodes as visited, so the recursion will visit some nodes multiple times, as drawn in the picture.

About the numeric precision on the backpropagation, i think that accumulating the gradient can be the best solution, but it is just my intuition.

If you want to add unit test, we can use Pytorch as a baseline.

**Comment by Leonardo16AM** [comment]
About the GPU Optimization, you have green light to do what you think is optimal :) . We can discuss possible improvements or solutions.

 So, i will close this PR, ok?

**Comment by DevNullx64** [comment]
Currently I'm fighting to split well value without gradiant, value with gradient and operation with gradiant.
I think go on use the vectorisation unit first. Before break again my alone neurone.

I'm turn around few moments then I will be back ><

**Comment by DevNullx64** [comment]
> About the GPU Optimization, you have green light to do what you think is optimal :) . We can discuss possible improvements or solutions.
> 
> So, i will close this PR, ok?

Sure, close it.
